lighthouse-core/audits/redirects.js

```diff
@@ -0,0 +1,94 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit');
+const Util = require('../report/v2/renderer/util');
+
+// PSI allows one redirect (http://example.com => http://m.example.com)
+const REDIRECT_THRESHOLD = 1;
+
+class Redirects extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'redirects',
+      description: 'Avoids page redirects.',
+      failureDescription: 'Has page redirects.',
+      helpText: ' Redirects introduce additional delays before the page can be loaded. [Learn more](https://developers.google.com/speed/docs/insights/AvoidRedirects).',
+      requiredArtifacts: ['URL', 'devtoolsLogs'],
+    };
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    return artifacts.requestCriticalRequestChains(artifacts.devtoolsLogs[Audit.DEFAULT_PASS])
+      .then(criticalRequests => {
+        const pageRedirects = [];
+        let totalWastedMs = 0;
+        let debugString = null;
+
+        let requests = criticalRequests;
+        let requestKey;
+        while (
+          (requestKey = Redirects.getKeyOfFirstRequest(requests)) &&
+          Redirects.isRedirect(requestKey)
+        ) {
+          const child = requests[requestKey];
+          const wastedMs = (child.request.endTime - child.request.startTime) * 1000;
+          totalWastedMs += wastedMs;
+          pageRedirects.push({
+            url: child.request.url,
+            wastedMs: Util.formatMilliseconds(wastedMs),
+          });
+
+          requests = child.children;
+        }
+
+        const passed = pageRedirects.length <= REDIRECT_THRESHOLD;
+        if (!passed) {
+          debugString = `Your page has ${pageRedirects.length} redirects.`;
+        }
+
+        const headings = [
+          {key: 'url', itemType: 'text', text: 'URL'},
+          {key: 'wastedMs', itemType: 'text', text: 'Wasted ms'},
+        ];
+        const details = Audit.makeTableDetails(headings, pageRedirects);
+
+        return {
+          debugString,
+          score: pageRedirects.length,
+          rawValue: totalWastedMs,
+          displayValue: Util.formatMilliseconds(totalWastedMs),
+          extendedInfo: {
+            value: {
+              wastedMs: totalWastedMs,
+            },
+          },
+          details,
+        };
+      });
+  }
+
+  static isRedirect(requestKey) {
+    return requestKey.includes('redirected');
```

Comment:
it looks like we depend here on 'redirected' being used in the name of request ID. IMO it'd be better to use something more solid like`WebInspector.NetworkRequest.redirects` (that'd require to change current logic quite a bit) or request status code.

Since I'm dealing with a similar thing here: https://github.com/GoogleChrome/lighthouse/pull/3311 . How about creating a `MainResource` computed artifact that exposes `WebInspector.NetworkRequest.redirects` that we both could use?

<img width="265" alt="screen shot 2017-09-21 at 17 28 27" src="https://user-images.githubusercontent.com/985504/30704222-5e17a40c-9ef2-11e7-9926-7459144e4c8e.png">

cc @brendankenny

Reply:
@kdzwinel Should we name it more like getMainResourceChain?

Reply:
@kdzwinel I think we have a slightly different use case :) I need to get the redirects as well so I cannot use the computed artifact

Reply:
@wardpeet maybe I'm missing something, but when you request a MainResource artifact you also get all NetworkRequest (that contain all timing info) representing all redirects that led to that resource. Check out this:

![](https://user-images.githubusercontent.com/985504/30787309-da0e2ac8-a185-11e7-9bbe-b8a51d65be3d.png)

Reply:
ah sorry I missed that! Great!

Reply:
Awesome 🙌  I already got one 👍 on my PR, so I hope to merge soon.

Reply:
MainResource computed artifact sgtm, we'll prob want to move away from more "chain-like" computed artifacts though going forward and try to make use of the `page-dependency-graph` instead